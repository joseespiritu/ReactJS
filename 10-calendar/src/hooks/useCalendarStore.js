import { useSelector, useDispatch } from "react-redux";
import { onAddNewEvent, onDeleteEvent, onSetActiveEvent, onUpdateEvent } from "../store/calendar/calendarSlice";

export const useCalendarStore = () => {

    const dispatch = useDispatch();

    const { events, activeEvent } = useSelector(state => state.calendar);

    const setActiveEvent = (calendarEvent) => {
        dispatch(onSetActiveEvent(calendarEvent));
    };

    const startSavingEvent = async (calendarEvent) => {
        // TODO: llegar al backend

        // todo bien
        if (calendarEvent._id) {
            // Actualizando
            dispatch(onUpdateEvent({ ...calendarEvent }));
        } else {
            // Creando
            dispatch(onAddNewEvent({ ...calendarEvent, _id: new Date().getTime() }));
        }
    };

    const startDeletingEvent = () => {
        // TODO: llegar al backend
        dispatch(onDeleteEvent());
    }

    return {
        // *Propiedades
        activeEvent,
        events,
        hasEventSelected: !!activeEvent,
        // *Metodos
        setActiveEvent,
        startSavingEvent,
        startDeletingEvent
    }
}