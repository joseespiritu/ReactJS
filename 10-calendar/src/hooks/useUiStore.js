import { useSelector, useDispatch } from "react-redux";
import { onCloseDateModal, onOpenDateModal } from "../store";

export const useUiStore = () => {
    const dispatch = useDispatch();
    const { isDateModalOpen } = useSelector(state => state.ui);

    const openDateModal = () => {
        dispatch(onOpenDateModal());
    };

    const closeDateModal = () => {
        dispatch(onCloseDateModal());
    };

    //* Tambien se puede implementar
    const toggleDateModal = () => {
        (isDateModalOpen)
            ? closeDateModal()
            : openDateModal();
    };

    return {
        //* Propiedades
        isDateModalOpen,
        //* Metodos
        openDateModal,
        closeDateModal,
    }
};