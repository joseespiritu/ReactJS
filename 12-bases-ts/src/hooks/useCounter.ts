import { useState, useRef, useEffect, useLayoutEffect } from 'react';
import { gsap } from 'gsap';


export const useCounter = ({maxCount = 1}) => {

    const [counter, setCounter] = useState(5);
    const elementToAnimate = useRef<any>(null);

    const tl = useRef(gsap.timeline());

    const handleClick = () => {
        // if (counter >= maxCount) return setCounter(counter);
        setCounter(prev => Math.min(prev + 1, maxCount));
    };

    useLayoutEffect(() => {

        if(!elementToAnimate.current) return;

        tl.current.to(elementToAnimate.current, {y: -10, duration: 0.2, ease: 'ease.out'})
                    .to(elementToAnimate.current, {y: 0, duration: 1, ease: 'bounce.out'})
                    .pause();
        
    }, []);

    useEffect(() => {
        // if ( counter < 10 ) return;
        // console.log('%cse llego al valor maximo', 'color: red; background-color: black;');
        // if(counter < maxCount) return;
        tl.current.play(0);
    }, [counter]);

    return {
        counter,
        elementToAnimate,
        handleClick,
    }
}