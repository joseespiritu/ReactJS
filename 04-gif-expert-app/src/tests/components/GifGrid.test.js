import React from "react";
import "@testing-library/jest-dom";
import { shallow } from "enzyme";
import GifGrid from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";
jest.mock("../../hooks/useFetchGifs");

describe("Pruebas en <GifGrid />", () => {
  const category = "One Punch";

  test("Debe de renderizar adecuadamente", () => {
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true,
    });
    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe de mostrar items cuando se cargan imagenes useFetchGifs", () => {
    const gifs = [
      {
        id: "ABC",
        url: "https:localhost/cualquiercosa",
        title: "cualquier cosa",
      },
    ];

    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false,
    });
    const wrapper = shallow(<GifGrid category={category} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("p").exists()).toBe(false);
    expect(wrapper.find("GifGridItem").length).toBe(gifs.length);
  });
});
