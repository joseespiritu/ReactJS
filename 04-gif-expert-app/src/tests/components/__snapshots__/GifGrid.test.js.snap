// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGrid /> Debe de mostrar items cuando se cargan imagenes useFetchGifs 1`] = `
<Fragment>
  <h3
    className="card animate__animated animate__fadeIn"
  >
    One Punch
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="cualquier cosa"
      url="https:localhost/cualquiercosa"
    />
  </div>
</Fragment>
`;

exports[`Pruebas en <GifGrid /> Debe de renderizar adecuadamente 1`] = `
<Fragment>
  <h3
    className="card animate__animated animate__fadeIn"
  >
    One Punch
  </h3>
  <p
    className="card animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
